File: src/main/java/org/zwobble/mammoth/util/Casts.java
package org.zwobble.mammoth.util;

import java.util.Optional;

public class Casts {
    public static <T> Optional<T> tryCast(Class<T> type, Object value) {
        try {
            return Optional.of(type.cast(value));
        } catch (ClassCastException exception) {
            return Optional.empty();
        }
    }
}

File: dotnet/Mammoth/Couscous/mammoth/internal/util/Casts.cs
using Mammoth.Couscous.java.util;

namespace Mammoth.Couscous.org.zwobble.mammoth.@internal.util {
    internal static class Casts {
        internal static Optional<T> tryCast<T>(System.Type type, object value) {
            if (value is T) {
                return new Some<T>((T) value);
            } else {
                return new None<T>();
            }
        }
    }
}
